In [1]:
!pip install sklearn_crfsuite
!pip install eli5

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 993.5/993.5 kB 26.7 MB/s eta 0:00:00
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 216.2/216.2 kB 9.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for eli5: filename=eli5-0.13.0-py2.py3-none-any.whl size=107730 sha256=067cc72a95577b4974a9bb2a1160c340bd9855c3406c306b9279852d48544d13
  Stored in directory: /root/.cache/pip/wheels/b8/58/ef/2cf4c306898c2338d51540e0922c8e0d6028e07007085c0004
Successfully built eli5


In [2]:
from build_features import text_to_features

# function to create features

def text2num(wrd):

  return[text_to_features(wrd, i) for i in range(len(wrd))]

# function to create labels

def text2lbl (wrd):

   return [label for token, label in wrd]

# Let's prepare the training data features using sent2features function which in turn uses text to features function.

X = [text2num(x) for x in sentences]

# Prepare the training data labels using text2lbl because the sentences are in a tuple consisting of words and tags.

y = [text2lb1(x) for x in sentences]

# Similarly, prepare the data.

test_X = [text2num(x) for x in sentences]

test_y = [text2lbl(x) for x in sentences]

# Now that the training and test data are ready, let's initialize the model. First, let's initialize and build a CRF model without hyperparameter tuning.

# building the CRF model

ner_crf_model = CRF(algorithm='lbfgs', max_iterations=25)

# tgraining the model with cross validation of 10

ner_predictions = cross_val_predict(estimator= ner_crf_model, X=X, y=y, cv=10)

#Let's evaluate the model on train data.

Accu_rpt = flat_classification_report(y_true=y,y_pred=ner_predictions,zero_division=1)

print(Accu_rpt)

                  precision    recall  f1-score   support

         B-Actor       0.94      0.92      0.93      5010
         B-Award       0.71      0.63      0.67       309
B-Character_Name       0.72      0.39      0.51      1024
      B-Director       0.85      0.81      0.83      1787
         B-Genre       0.84      0.80      0.82      3384
       B-Opinion       0.48      0.38      0.42       810
        B-Origin       0.52      0.42      0.47       779
          B-Plot       0.49      0.47      0.48      6468
         B-Quote       0.78      0.37      0.51       126
  B-Relationship       0.68      0.51      0.58       580
    B-Soundtrack       0.69      0.22      0.33        50
          B-Year       0.96      0.97      0.97      2702
         I-Actor       0.94      0.92      0.93      6121
         I-Award       0.81      0.73      0.77       719
I-Character_Name       0.71      0.40      0.51       760
      I-Director       0.89      0.81      0.85      1653
         I-Ge

In [3]:
# building the CRF model

crf_ner = CRF(algorithm='lbfgs', max_iterations=25)

#Fitting model on train data.

crf_ner.fit(X,y)

# prediction on test data

test_prediction=crf_ner.predict(test_X)

# get labels

lbs=list(crf_ner.classes_)

#get accuracy

metrics.flat_f1_score(test_y, test_prediction, average= 'weighted', labels=lbs)

0.8369950502328535


In [4]:
#sort the labels

sorted_lbs=sorted(lbs, key= lambda name:(name [1:], name[0]))

#get classification report

print(metrics.flat_classification_report(test_y, test_prediction, labels=sorted_lbs, digits=4))

                    precision    recall  f1-score  support

               O       0.8714    0.8423    0.8566    14143
         B-Actor       0.9265    0.9207    0.9236     1274
         I-Actor       0.9336    0.9227    0.9281     1553
         B-Award       0.6923    0.6818    0.6870       66
         I-Award       0.7941    0.7347    0.7633      147
B-Character_Name       0.7283    0.4452    0.5526      283
I-Character_Name       0.7453    0.5286    0.6186      227
      B-Director       0.8575    0.8494    0.8534      425
      I-Director       0.9116    0.8783    0.8947      411
         B-Genre       0.8355    0.8048    0.8199      789
         I-Genre       0.7853    0.7463    0.7653      544
       B-Opinion       0.4867    0.3744    0.4232      195
       I-Opinion       0.3333    0.1608    0.2170      143
        B-Origin       0.4803    0.3842    0.4269      190
        I-Origin       0.7088    0.6386    0.6719      808
          B-Plot       0.4859    0.4705    0.4781     1

The test data result shows an overall F1 score of 0.8370 and an accuracy of 0.8412. The accuracy increased by nearly 40% by adding more features, and the F1 score increased by 0.5. We can further increase the accuracy by performing hyperparameter tuning.

Let's take a random sentence and predict the tag using the trained model.

1. Create feature maps, as we did for training data, for input sentences using the word2feature function.

2. Convert into an array and predict using crf.predict(input_vector).

Next, convert each word into features.


In [5]:
def text_2_ftr_sample(words):
  return [text_to_features(words, i) for i in range(len(words))]

# Split the sentences and convert each word into features.

# define sample sentence

X_sample=['alien invasion is the movie directed by christoper nollen'.split()]

# convert to features

X_sample=[text_2_ftr_sample(x) for x in X_sample]

# predicting the class

crf_ner.predict(X_sample1)

[['B-Actor', 'I-Actor', 'O', 'O', 'O', 'O', 'O', 'B-Director','I-Director']] 


Now, let's try the BERT model and check if it performs better than the CRF model.

# BERT Transformer

BERT (Bidirectional Encoder Representations from Transformers) is a model that is trained on large data sets. This pretrained model can be fine-tuned as per the requirement and used for different tasks such as sentiment analysis, question answering system, sentence classification, and others. BERT transfers learning in NLP, and it is a state-of-the-art method.

BERT uses transformers, mainly the encoder part. The attention mechanism learns the contextual relationship between words and subwords. Unlike other models, Transformer's encoder learns all sequences at once. The input is a sequence of words (tokens) that are embed into vectors and then proceed to the neural networks. The output is the sequence of tokens that corresponds to the input token of the given sequence.

Let's implement the BERT model.


In [6]:
# importing necessary libraries

from sklearn.model_selection import train_test_split

from sklearn.metrics import accuracy_score

from sklearn.metrics import classification_report, make_scorer

# importing NER models from simple transformers

!pip install simpletransformers

from simpletransformers.ner import NERModel, NERArgs

# importing libraries for evaluation

from sklearn_crfsuite.metrics import flat_classification_report

from sklearn_crfsuite import CRF, scorers, metrics

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 250.7/250.7 kB 9.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.2/7.2 MB 86.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 486.2/486.2 kB 43.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 4.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 73.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 57.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.9/8.9 MB 70.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 50.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 236.8/236.8 kB 25.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 77.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 184.3/184.3 kB 16.0

In [7]:
from sklearn.preprocessing import LabelEncoder
# Let's encode the sentence column using Label Encoder.

# encoding sentence values

data["sentence_id"] = LabelEncoder().fit_transform(data["sentence_id"])

data1["sentence_id"] = LabelEncoder().fit_transform(data1["sentence_id"])

# Let's convert all labels into uppercase.

#converting labels to upper string as it is required format

data["labels"] = data["labels"].str.upper()

data1["labels"] = data["labels"].str.upper()

# Next, separate the train and test data

X= data[["sentence_id", "words"]]

Y =data["labels"]

# Then, create a train and test data frame.

# building up train and test data to dataframe

ner_tr_dt = pd.DataFrame({"sentence_id": data["sentence_id"], "words":data["words"],"labels":data["labels"]})

test_data = pd.DataFrame({"sentence_id": data1["sentence_id"], "words":data1["words"],"labels":data1 ["labels"]})

# Also, let's store the list of unique labels.

#label values

label = ner_tr_dt["labels"].unique().tolist()

We need to fine-tune the BERT model so that we can use parameters. Here, we changed epochs number and batch size.

To improve the model further, we can change other parameters as well.


In [8]:
#fine tuning our model on custom data

args = NERArgs()

#set the # of epoch

args.num_train_epochs = 2

#learning rate

args.learning_rate = 1e-6

args.overwrite_output_dir =True

#train and evaluation batch size

args.train_batch_size = 6

args.eval_batch_size = 6

In [9]:
#initializing the model

Ner_bert_mdl= NERModel('bert', 'bert-base-cased', labels=label, args = args)

#training our model

Ner_bert_mdl.train_model(ner_tr_dt, eval_data = test_data, acc=accuracy_score)

100%|███████████████████████| 3/3 [00:08<00:00, 8.78it/s]

Epoch 5 of 5: 100%|███████████████████████| 5/5 [07:38<00:00, 95.51it/s]

Epochs 0/5. Running Loss: 0.4331: 100%|███████████████████████| 489/489 [01:20<00:00, 6.66it/s]

Epochs 1/5. Running Loss: 0.4114: 100%|███████████████████████| 489/489 [01:21<00:00, 6.61it/s]

Epochs 2/5. Running Loss: 0.2505: 100%|███████████████████████| 489/489 [01:21<00:00, 6.61it/s]

Epochs 3/5. Running Loss: 0.1116: 100%|███████████████████████| 489/489 [01:21<00:00, 6.56it/s]

Epochs 4/5. Running Loss: 0.0743: 100%|███████████████████████| 489/489 [01:45<00:00, 6.60it/s]

(2445, 0.27327230073252584)


We can observe a loss of 0.27 in the final epoch

In [10]:
# function to store labels and words of each sentence in list

class sent_generate(object):

  def __init__(self, data):

    self.n_sent = 1.0

    self.data = data

    self.empty = False

    fn_group = lambda s: [(a, b) for a,b in zip(s["words"].values.tolist(),s["labels"].values.tolist())]

    self.grouped = self.data.groupby("sentence_id").apply(fn_group)

    self.sentences = [x for x in self.grouped]

# storing words and labels of each sentence in single list of train data

Sent_get = sent_generate(ner_tr_dt)

sentences = Sent_get.sentences

print(sentences[0])

# This is how a sentence will look like.

def txt_2_lbs (sent):

  return [label for token, label in sent]

y_train_group = [txt_2_lbs(x) for x in sentences[:4]]

 [('steve', 'B-ACTOR'), ('mcqueen', 'I-ACTOR'), ('provided', 'O'), ('a', 'O'), ('thrilling', '8-PLOT')]


In [11]:
# Storing words and labels of each sentence in single list of test data

Sent_get = sent_generate(test_data)

Sentences = Sent_get.sentences

#This is how a sentence will look like.

print(sentences[0])

def txt_2_lbs(sent):

  return [label for token, label in sent]

y_test = [txt_2_lbs(x) for x in sentences[:5]]

[('1', 'O'), ('need', 'O'), ('that', 'O'), ('movie', 'O'), ('which', 'O')]


In [12]:
# Let's evaluate the model on test data.

#evaluating on test data

result, model_outputs, preds_list = Ner_bert_mdl.eval_model(test_data)

#individual group report

accu_rpt = flat_classification_report(y_pred=preds_list, y_true=y_test)

print(accu_rpt)

                  precision    recall  f1-score   support

         B-Actor       0.95      0.97      0.96      1274
         B-Award       0.68      0.70      0.69        66
B-Character_Name       0.71      0.72      0.72       283
      B-Director       0.88      0.92      0.90       425
         B-Genre       0.83      0.85      0.84       789
       B-Opinion       0.47      0.52      0.50       195
        B-Origin       0.48      0.45      0.47       190
          B-Plot       0.54      0.52      0.53      1577
         B-Quote       0.86      0.81      0.84        47
  B-Relationship       0.72      0.67      0.69       171
    B-Soundtrack       0.44      0.50      0.47         8
          B-Year       0.98      0.98      0.98       661
         I-Actor       0.96      0.96      0.96      1553
         I-Award       0.77      0.81      0.79       147
I-Character_Name       0.69      0.68      0.69       227
      I-Director       0.95      0.93      0.94       411
         I-Ge

The accuracy of test data is 88%. Let's evaluate the model on train data also to see if there is any overfitting.


In [18]:
#evaluating on train data

result_train, model_outputs_train, preds_list_train = Ner_bert_mdl.eval_model(ner_tr_dt)

#individual group report of train data

report_train = flat_classification_report (y_pred=preds_list, y_true = y_train_group)

print(report_train)

                  precision    recall  f1-score   support

         B-Actor       1.00      1.00      1.00      5010
         B-Award       0.97      0.97      0.97       309
B-Character_Name       0.99      0.99      0.99      1024
      B-Director       0.94      0.99      0.97      1787
         B-Genre       0.97      0.95      0.96      3384
       B-Opinion       0.83      0.94      0.88       810
        B-Origin       0.88      0.86      0.87       779
          B-Plot       0.94      0.93      0.94      6468
         B-Quote       0.99      0.95      0.97       126
  B-Relationship       0.91      0.90      0.90       580
    B-Soundtrack       0.92      0.92      0.92        50
          B-Year       1.00      1.00      1.00      2702
         I-Actor       1.00      1.00      1.00      6121
         I-Award       0.96      0.99      0.97       719
I-Character_Name       0.99      0.99      0.99       760
      I-Director       1.00      1.00      1.00      1653
         I-Ge

The accuracy of the training data is 98%. Compared to the CRF model, the BERT model is performing great.

Now, let's take a random sentence and predict the tag using the BERT model built.

In [19]:
prediction, model_output = Ner_bert_mdl.predict(["aliens invading is movie by christoper nollen"])

# Here, the model generates predictions that tag given words from each sentence, and model_output are outputs generated by the model.

[[{'aliens': 'B-PLOT'}, 
  {'invading': 'I-PLOT'}, 
  {'is': '0'}, 
  {'movie': '0'}, 
  {'by': '0'}, 
  {'christoper': 'B-DIRECTOR'}, 
  {'nollen': 'I-DIRECTOR'}]]


# Next Steps

*   Add more features to the model example combination of words with gives a proper meaning for CRF model.

*   The current implementation considers only two hyperparameters in the CV search for the CRF model; however, the CRF model offers more parameters that can be further tuned to improve the performance.

*  Use an LSTM neural network and creating an ensemble model with CRF

*  For the BERT model, we can add a CRF layer to get abhore efficient network as BERT finds efficient patterns between words.


## Named Entity Recognition Using CRF and BERT ends here.